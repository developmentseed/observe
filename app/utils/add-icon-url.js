import * as taginfo from '../presets/taginfo.json'
import { uninterestingKeys } from './uninterestingKeys'
import _omit from 'lodash.omit'
import _find from 'lodash.find'

/**
 * Parses through each feature, and adds icon url to properties
 * The icon property is used for styling later
 * @param {Object} geojson
 * @return {Object} geojson
 */
export function addIconUrl (geojson) {
  if (geojson.type === 'FeatureCollection') {
    const points = geojson.features.filter(feature => {
      return feature.geometry.type === 'Point'
    })
    points.forEach(feature => {
      setIcon(feature)
    })
  } else {
    setIcon(geojson)
  }
  return geojson
}

function setIcon (feature) {
  feature.properties.icon = undefined
  const featureTags = _omit(feature.properties, uninterestingKeys)
  const keys = Object.keys(featureTags)
  for (const index in keys) {
    const key = keys[index]
    const match = _find(taginfo.tags, { 'key': key, 'value': featureTags[key] })
    if (match) {
      if (match.icon_url != null) {
        feature.properties.icon = match.icon_url.replace(/-/g, '_')
      }
      break
    }
  }
  if (!feature.properties.icon || feature.properties.icon.startsWith('iD-')) {
    feature.properties.icon = 'maki_marker'
  }
  return feature
}
