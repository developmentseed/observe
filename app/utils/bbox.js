import tileCover from '@mapbox/tile-cover'
import bboxPolygon from '@turf/bbox-polygon'
import turfBbox from '@turf/bbox'
import turfBuffer from '@turf/buffer'
import _flatten from 'lodash.flatten'

// TODO - move this to an appropriate config file
const tileLimits = {
  min_zoom: 16,
  max_zoom: 16
}

/**
 * Convert map bbox object to a string
 * @param {*} bbox
 * @return string
 */
export function bboxToString (bbox) {
  const ne = bbox[0]
  const sw = bbox[1]
  const left = sw[0]
  const bottom = sw[1]
  const right = ne[0]
  const top = ne[1]
  return [left, bottom, right, top].join(',')
}

export function bboxToTiles (bbox) {
  const polygon = bboxPolygon(_flatten(bbox))
  const tiles = tileCover.indexes(polygon.geometry, tileLimits)
  return tiles
}

export function featureToTiles (feature) {
  let bbox = []
  if (feature.geometry && feature.geometry.type === 'Point') {
    const buffer = turfBuffer(feature, '0.0005', { 'units': 'kilometers' })
    bbox = turfBbox(buffer)
  } else {
    bbox = turfBbox(feature)
  }
  const polygon = bboxPolygon(bbox)
  const tiles = tileCover.indexes(polygon.geometry, tileLimits)
  return tiles
}
