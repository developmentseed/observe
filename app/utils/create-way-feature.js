import getRandomId from '../utils/get-random-id'

export default function createWayFeature (nodes = [], properties = {}, options = {}) {
  if (!properties.ndrefs) {
    properties.ndrefs = []
  }

  if (!options.id) {
    options.id = getRandomId()
  }
  properties.id = options.id

  const coordinates = nodes.map((node, index) => {
    // populate ndrefs
    properties.ndrefs.push(node.properties.id)
    if (!node.properties.ways) node.properties.ways = {}
    if (!node.properties.ways.hasOwnProperty(properties.id)) {
      node.properties.ways[properties.id] = index
    }
    return node.geometry.coordinates
  })

  properties.id = `way/${options.id}`
  return {
    type: 'Feature',
    id: `way/${options.id}`,
    properties,
    geometry: {
      type: 'LineString',
      coordinates
    }
  }
}
