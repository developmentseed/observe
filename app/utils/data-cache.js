import LRU from 'lru-cache'
import sizeof from 'object-sizeof'

export default new LRU({
  length: (val, key) => sizeof(val),
  max: 5e6, // 5MB
  maxAge: 3600e3 // 1 hour
})
