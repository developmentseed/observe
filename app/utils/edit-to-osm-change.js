import { DOMParser } from 'xmldom'
import _omit from 'lodash.omit'
import { nonpropKeys } from '../utils/uninterestingKeys'
import _get from 'lodash.get'
import _uniq from 'lodash.uniq'
import { isInvalidFeature } from '../utils/utils'

/**
 * Takes an edit and transforms it into `osmChange` XML to be uploaded to the OSM API
 * @param {Object} edit - an "edit" item, containing `type`, `oldFeature` and `newFeature`
 * @param {Number} changesetId - changeset id
 * @param {FeatureCollection} memberNodes - feature collection of member nodes (needed when deleting a way)
 * @return {String<XML>} - osmChange XML string
 */
export default function (edit, changesetId, memberNodes = null) {
  console.log('edit', edit)
  const isSimpleChange = !(edit.newFeature && edit.newFeature.wayEditingHistory)
  if (isSimpleChange) {
    return {
      osmChangeXML: getSimpleChange(edit, changesetId, memberNodes),
      nodeIdMap: null
    }
  } else {
    return getComplexChange(edit, changesetId)
  }
}

/**
 * @param {Object<GeoJSON Feature>} feature
 * @returns {Object} - key value pairs of osm tags for feature (removed uninteresting tags)
 */
function getTags (feature) {
  // FIXME: get from app consts
  const uninterestingProps = [
    'id',
    'version',
    'user',
    'uid',
    'changeset',
    'timestamp',
    'icon',
    'edge',
    'distance',
    'ways',
    'ndrefs',
    'addedNodes',
    'movedNodes',
    'deletedNodes',
    'modifiedNodes',
    'mergedNodes',
    'membership',
    'dist',
    'index',
    'location'
  ]
  return _omit(feature.properties, uninterestingProps)
}

/**
 *
 * @param {Object} edit - containing newFeature.wayEditingHistory with details of the complex edit
 * @param {Number} changesetId
 */
function getComplexChange (edit, changesetId) {
  console.log('called get complex change')
  let negativeIdCounter = 0

  function getNextNegativeId () {
    negativeIdCounter = negativeIdCounter - 1
    return negativeIdCounter
  }

  // const changes = edit.type === 'create' ? getComplexCreate(edit) : getComplexModify(edit)
  // const { creates, modifies, deletes } = changes
  let creates = []

  let modifies = []

  let deletes = []

  const { wayEditingHistory, ...feature } = edit.newFeature

  feature.properties = { ...edit.newFeature.properties }

  if (feature.properties.id.split('/').length > 1) {
    feature.properties.id = feature.properties.id.split('/')[1]
  }

  const nodeIdMap = wayEditingHistory.way.nodes.reduce((mapping, node, idx) => {
    const id = node.properties.id
    if (isNewId(id)) {
      mapping[id] = getNextNegativeId()
    } else if (id.split('/').length > 1) {
      mapping[id] = id.split('/')[1]
    } else {
      mapping[id] = id
    }
    return mapping
  }, {})

  // replace ndrefs of temporary ids with negative ids generated in mapping above
  feature.properties.ndrefs = feature.properties.ndrefs.map(ref => {
    if (nodeIdMap.hasOwnProperty(ref)) {
      return nodeIdMap[ref]
    } else {
      return ref
    }
  })

  console.log('feature ndrefs', feature.properties.ndrefs)
  wayEditingHistory.addedNodes.forEach(addedNodeId => {
    console.log('added node id', addedNodeId)
    const node = wayEditingHistory.way.nodes.find(nd => nd.properties.id === addedNodeId)
    console.log('node', node)
    const id = node.properties.id
    if (isNewId(id) && !creates.find(create => create.id === nodeIdMap[id])) {
      creates.push({
        type: 'node',
        id: nodeIdMap[id],
        feature: node
      })
    }
  })

  wayEditingHistory.movedNodes.forEach(movedNodeId => {
    const node = wayEditingHistory.way.nodes.find(nd => nd.properties.id === movedNodeId)
    const id = node.properties.id
    modifies.push({
      type: 'node',
      id: nodeIdMap[id],
      feature: node
    })
  })

  wayEditingHistory.deletedNodes.forEach(deletedNodeId => {
    const node = wayEditingHistory.way.nodes.find(nd => nd.properties.id === deletedNodeId)
    const id = node.properties.id
    deletes.push({
      type: 'node',
      id: nodeIdMap[id],
      feature: node
    })
  })

  wayEditingHistory.mergedNodes.forEach(mergedNode => {
    const sourceNodeId = mergedNode.sourceNode
    const node = wayEditingHistory.way.nodes.find(nd => nd.properties.id === sourceNodeId)
    const id = node.properties.id
    const nodeTags = getTags(node)
    if (Object.keys(nodeTags).length === 0) { // node had no other tags, can be deleted
      deletes.push({
        type: 'node',
        id: nodeIdMap[id],
        feature: node
      })
    }
  })

  wayEditingHistory.modifiedSharedWays.forEach(way => {
    if (way.properties.id.split('/').length > 1) {
      way.properties.id = way.properties.id.split('/')[1]
    }
    // if shared way is same as top level feature, ignore
    if (way.properties.id !== feature.properties.id) {
      // if way has only moved nodes, ignore
      if (_get(way, 'properties.addedNodes', []).length > 0 || _get(way, 'properties.deletedNodes', []).length > 0) {
        // for nodes with new ids, retrieve the current negative id mapping
        way.properties.ndrefs = way.properties.ndrefs.map(ndref => {
          if (isNewId(ndref)) {
            return nodeIdMap[ndref]
          } else {
            return ndref
          }
        })

        if (isInvalidFeature(way)) {
          // if the way now contains less than 2 nodes (3 for polygons), delete it
          deletes.push({
            type: 'way',
            id: way.properties.id,
            feature: way
          })
          const ndrefs = _uniq(way.properties.ndrefs)
          ndrefs.forEach(nd => {
            const node = wayEditingHistory.way.nodes.find(n => n.properties.id === `node/${nd}`)
            deletes.push({
              type: 'node',
              id: nodeIdMap[node.properties.id],
              feature: node
            })
          })
        } else {
          // in the normal case, push a modify operation
          modifies.push({
            type: 'way',
            id: way.properties.id,
            feature: way
          })
        }
      }
    }
  })

  // make sure we add the XML for the main feature last to make sure
  // all references have been added before
  if (edit.type === 'create') {
    creates.push({
      type: 'way',
      id: getNextNegativeId(),
      feature
    })
  } else if (edit.type === 'modify') {
    if (isInvalidFeature(feature)) {
      // way does not contain enough nodes, delete it.
      deletes.push({
        type: 'way',
        id: feature.properties.id,
        feature
      })
      const ndrefs = _uniq(feature.properties.ndrefs)
      ndrefs.forEach(nd => {
        const node = wayEditingHistory.way.nodes.find(n => n.properties.id === `node/${nd}`)
        deletes.push({
          type: 'node',
          id: nodeIdMap[node.properties.id],
          feature: node
        })
      })
    } else if (wayEditingHistory.addedNodes.length > 0 || wayEditingHistory.deletedNodes.length > 0 || wayEditingHistory.mergedNodes.length > 0) {
      // if feature is modified, we only need to include in change XML if nodes
      // were added or removed
      modifies.push({
        type: 'way',
        id: feature.properties.id,
        feature
      })
    } else {
      // if way has only moved nodes, do nothing
    }
  }

  const XMLForChanges = getXMLForChanges({ creates, modifies, deletes }, changesetId)
  return {
    osmChangeXML: XMLForChanges,
    nodeIdMap: nodeIdMap
  }
}

function getXMLForChanges ({ creates, modifies, deletes }, changesetId) {
  const xmlRoot = '<osmChange></osmChange>'
  const parser = new DOMParser()
  const xmlDoc = parser.parseFromString(xmlRoot, 'text/xml')
  const rootElem = xmlDoc.getElementsByTagName('osmChange')
  if (creates.length > 0) {
    const createNode = xmlDoc.createElement('create')
    creates.forEach(create => {
      const featureType = create.type
      const feature = create.feature
      const id = create.id
      const elemNode = xmlDoc.createElement(featureType)
      elemNode.setAttribute('id', id) // TODO: Verify id is consistent
      elemNode.setAttribute('changeset', changesetId)
      elemNode.setAttribute('version', 1) // QUESTION: is this correct for new features?
      if (featureType === 'node') {
        elemNode.setAttribute('lon', feature.geometry.coordinates[0])
        elemNode.setAttribute('lat', feature.geometry.coordinates[1])
      }
      const tags = getTags(feature)
      addTags(xmlDoc, elemNode, tags)

      if (featureType === 'way') {
        addNdrefs(xmlDoc, elemNode, feature.properties.ndrefs)
      }
      createNode.appendChild(elemNode)
    })
    rootElem[0].appendChild(createNode)
  }
  if (modifies.length > 0) {
    const modifyNode = xmlDoc.createElement('modify')
    modifies.forEach(modify => {
      const featureType = modify.type
      const feature = modify.feature
      const id = modify.id
      const elemNode = xmlDoc.createElement(featureType)
      elemNode.setAttribute('id', id)
      elemNode.setAttribute('changeset', changesetId)
      elemNode.setAttribute('version', feature.properties.version)
      if (featureType === 'node') {
        elemNode.setAttribute('lon', feature.geometry.coordinates[0])
        elemNode.setAttribute('lat', feature.geometry.coordinates[1])
      }
      const tags = getTags(feature)
      addTags(xmlDoc, elemNode, tags)
      if (featureType === 'way') {
        addNdrefs(xmlDoc, elemNode, feature.properties.ndrefs)
      }
      modifyNode.appendChild(elemNode)
    })
    rootElem[0].appendChild(modifyNode)
  }
  if (deletes.length > 0) {
    const deleteNode = xmlDoc.createElement('delete')
    deleteNode.setAttribute('if-unused', 'true')
    deletes.forEach(del => {
      const featureType = del.type
      const feature = del.feature
      const id = del.id
      const elemNode = xmlDoc.createElement(featureType)
      elemNode.setAttribute('id', id)
      elemNode.setAttribute('changeset', changesetId)
      elemNode.setAttribute('version', feature.properties.version)
      if (featureType === 'node') {
        elemNode.setAttribute('lon', feature.geometry.coordinates[0])
        elemNode.setAttribute('lat', feature.geometry.coordinates[1])
      }
      const tags = getTags(feature)
      addTags(xmlDoc, elemNode, tags)
      if (featureType === 'way') {
        addNdrefs(xmlDoc, elemNode, feature.properties.ndrefs)
      }
      deleteNode.appendChild(elemNode)
    })
    rootElem[0].appendChild(deleteNode)
  }
  // console.log('xml', xmlDoc.toString())
  return xmlDoc.toString()
}

function addTags (xmlDoc, elem, tags) {
  Object.keys(tags).forEach(key => {
    const tagElem = xmlDoc.createElement('tag')
    tagElem.setAttribute('k', key)
    tagElem.setAttribute('v', tags[key])
    elem.appendChild(tagElem)
  })
  return elem
}

function addNdrefs (xmlDoc, elem, refs) {
  refs.forEach(ref => {
    ref = String(ref) // normalize to strings
    if (ref.split('/').length > 1) {
      ref = ref.split('/')[1]
    }
    const refElem = xmlDoc.createElement('nd')
    refElem.setAttribute('ref', ref)
    elem.appendChild(refElem)
  })
  return elem
}

function isNewId (id) {
  id = String(id)
  return id.startsWith('observe-')
}

/**
 * Function to get XML for "simple" changes that result in a single change operation
 * For the more complex case of way edits, where other ways may be affected, etc. we use
 * getComplexChange
 */
function getSimpleChange (edit, changesetId, memberNodes) {
  const xmlRoot = '<osmChange></osmChange>'
  const parser = new DOMParser()
  const xmlDoc = parser.parseFromString(xmlRoot, 'text/xml')
  const changeTypeNode = xmlDoc.createElement(edit.type)
  if (edit.type === 'delete') {
    changeTypeNode.setAttribute('if-unused', 'true')
  }
  const rootElem = xmlDoc.getElementsByTagName('osmChange')
  const feature = edit.type === 'delete' ? edit.oldFeature : edit.newFeature
  let [ featureType, id ] = feature.id.split('/')

  // for newly created elements, always set id to `-1`
  if (edit.type === 'create') {
    id = '-1'
  }

  const changeNode = xmlDoc.createElement(featureType)
  changeNode.setAttribute('id', id)
  changeNode.setAttribute('changeset', changesetId)
  changeNode.setAttribute('version', feature.properties.version)
  if (featureType === 'node') {
    changeNode.setAttribute('lon', feature.geometry.coordinates[0])
    changeNode.setAttribute('lat', feature.geometry.coordinates[1])
  }
  const tags = getTags(feature)
  addTags(xmlDoc, changeNode, tags)

  if (featureType === 'way') {
    const refs = feature.properties.ndrefs
    refs.forEach(ref => {
      const refElem = xmlDoc.createElement('nd')
      refElem.setAttribute('ref', ref)
      changeNode.appendChild(refElem)
    })
  }
  changeTypeNode.appendChild(changeNode)

  // If it is a way being deleted, add deletes for all orphaned member nodes
  if (edit.type === 'delete' && featureType === 'way') {
    if (!memberNodes) {
      throw new Error('deleting a way, but no member nodes supplied')
    }
    memberNodes.features.forEach(feature => {
      const osmProps = _omit(feature.properties, nonpropKeys)

      // Only delete the node if it doesn't have any osm tags
      if (Object.keys(osmProps).length === 0) {
        const nodeElem = xmlDoc.createElement('node')
        nodeElem.setAttribute('id', feature.properties.id.split('/')[1])
        nodeElem.setAttribute('lon', feature.geometry.coordinates[0])
        nodeElem.setAttribute('lat', feature.geometry.coordinates[1])
        nodeElem.setAttribute('version', feature.properties.version)
        nodeElem.setAttribute('changeset', changesetId)
        changeTypeNode.appendChild(nodeElem)
      }
    })
  }
  rootElem[0].appendChild(changeTypeNode)
  return xmlDoc.toString()
}
