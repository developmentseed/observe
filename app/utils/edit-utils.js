import { UploadNetworkError } from './errors'

/**
 * Given an edit, says if the edit can be retried based on the error type
 * @param {Object} edit
 * @returns {Boolean} - true if retryable, false if not
 */
export function isRetryable (edit) {
  return getError(edit).retryable
}

/**
 * For an array of edit objects, returns a filtered list of retriable edits
 * @param {Array<Object>} edits - array of edit objects
 * @returns {Array<Object>}
 */
export function getAllRetriable (edits) {
  return edits.filter(e => isRetryable(e))
}

/**
 * Given an edit, returns whether it has a Version Conflict error
 * @param {Object} edit
 * @returns {Boolean}
 */
export function isConflict (edit) {
  return getError(edit).code === 'VersionMismatchError'
}

/**
 * Return the applicable / current error for an edit.
 * If it has errors, return the last error in the error.
 * If it does not have errors, we assume that it is waiting for network, and return a NetworkError
 * @param {Object} edit - edit object
 * @return {ObserveError} - an instance of ObserveError, with the error for the edit
 */
export function getError (edit) {
  return edit.errors.length > 0 ? edit.errors.slice(-1)[0] : new UploadNetworkError()
}

export function getErrorExtra (edit) {
  return getError(edit).extra
}
