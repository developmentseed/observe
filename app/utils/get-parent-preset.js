import { presets } from '../presets/presets.json'

/**
 * Returns parent preset of preset
 * For example, returns amenity/bank for the amenity/bank/Citi Bank preset
 * Based on https://github.com/openstreetmap/iD/blob/master/modules/presets/preset.js#L11
 * @param {Object} preset
 * @return {Object} preset
 */

export function getParentPreset (preset) {
  if (!preset || !preset.key) return undefined

  const endIndex = preset.key.lastIndexOf('/')
  if (endIndex < 0) return null

  const parentId = preset.key.substring(0, endIndex)

  if (!presets[parentId]) return

  const parentPreset = { ...presets[parentId] }

  if (presets[parentId].fields && Array.isArray(presets[parentId].fields)) {
    parentPreset.fields = [...presets[parentId].fields]
  }
  return parentPreset
}
