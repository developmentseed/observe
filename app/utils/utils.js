export function isNewId (id) {
  id = String(id)
  return id.startsWith('observe-')
}
