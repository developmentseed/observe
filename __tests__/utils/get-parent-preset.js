/* global fetch */
/* global test */
/* global __dirname */
/* global expect */

import { getParentPreset } from '../../app/utils/get-parent-preset'

test('get parent preset', () => {
  const presetBank = { 'name': 'Agribank', 'icon': 'maki-bank', 'geometry': ['point', 'area'], 'tags': { 'amenity': 'bank', 'brand:wikidata': 'Q4693829' }, 'addTags': { 'amenity': 'bank', 'brand': 'Agribank', 'brand:wikidata': 'Q4693829', 'brand:wikipedia': 'en:AgriBank', 'name': 'Agribank' }, 'removeTags': { 'amenity': 'bank', 'brand': 'Agribank', 'brand:wikidata': 'Q4693829', 'brand:wikipedia': 'en:AgriBank', 'name': 'Agribank' }, 'matchScore': 2, 'suggestion': true, 'key': 'amenity/bank/Agribank' }

  const presetBuilding = { 'icon': 'maki-building', 'geometry': ['area'], 'tags': { 'building': 'apartments' }, 'matchScore': 0.5, 'name': 'Apartment Building', 'key': 'building/apartments' }
  const parentBankPreset = getParentPreset(presetBank)
  const parentBuildingPreset = getParentPreset(presetBuilding)
  expect(parentBankPreset).toMatchSnapshot()
  expect(parentBuildingPreset).toMatchSnapshot()
})
