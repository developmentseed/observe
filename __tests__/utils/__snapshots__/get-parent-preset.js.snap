// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get parent preset 1`] = `
Object {
  "fields": Array [
    "name",
    "operator",
    "address",
    "building_area",
    "atm",
    "drive_through",
  ],
  "geometry": Array [
    "point",
    "area",
  ],
  "icon": "maki-bank",
  "moreFields": Array [
    "opening_hours",
    "currency_multi",
    "website",
    "phone",
    "email",
    "fax",
    "wheelchair",
  ],
  "name": "Bank",
  "tags": Object {
    "amenity": "bank",
  },
  "terms": Array [
    "credit union",
    "check",
    "deposit",
    "fund",
    "investment",
    "repository",
    "reserve",
    "safe",
    "savings",
    "stock",
    "treasury",
    "trust",
    "vault",
  ],
}
`;

exports[`get parent preset 2`] = `
Object {
  "fields": Array [
    "name",
    "building",
    "levels",
    "height",
    "address",
  ],
  "geometry": Array [
    "point",
    "area",
  ],
  "icon": "maki-home",
  "matchScore": 0.6,
  "moreFields": Array [
    "architect",
    "building/material",
    "layer",
    "roof/colour",
    "smoking",
    "wheelchair",
  ],
  "name": "Building",
  "tags": Object {
    "building": "*",
  },
  "terms": Array [],
}
`;
